File: frameworks/motia/motia-py/packages/motia/src/motia/iii.py
"""III SDK client initialization for Motia framework."""

import os
from pathlib import Path
from typing import Any

try:
    import tomllib  # type: ignore[import-not-found]
except ImportError:
    import tomli as tomllib  # type: ignore[import-not-found]

from iii import IIIClient, register_worker
from iii.iii_constants import InitOptions, TelemetryOptions

_engine_ws_url = os.environ.get("III_URL", "ws://localhost:49134")
_instance: IIIClient | None = None


def _read_project_name() -> str | None:
    """Walk up from cwd (max 1 parent) to find the nearest pyproject.toml and extract the project name."""
    max_depth = 1
    directory = Path.cwd()
    for _ in range(max_depth + 1):
        pyproject = directory / "pyproject.toml"
        if pyproject.exists():
            try:
                with open(pyproject, "rb") as f:
                    data = tomllib.load(f)
                name = data.get("project", {}).get("name") or data.get("tool", {}).get("poetry", {}).get("name")
                if name:
                    return str(name)
            except Exception:
                pass
        parent = directory.parent
        if parent == directory:
            break
        directory = parent
    return None


def _create_iii(otel_config: dict[str, Any] | None = None) -> IIIClient:
    telemetry = TelemetryOptions(
        framework="motia",
        project_name=_read_project_name(),
    )
    return register_worker(_engine_ws_url, InitOptions(telemetry=telemetry, otel=otel_config))


def get_instance() -> IIIClient:
    """Get the III SDK singleton instance.

    Creates a default instance if none exists.
    """
    global _instance
    if _instance is None:
        _instance = _create_iii()
    return _instance


def init_iii(otel_config: dict[str, Any] | None = None) -> IIIClient:
    """Initialize the III SDK with optional OpenTelemetry configuration.

    Args:
        otel_config: Optional OpenTelemetry configuration dict.

    Returns:
        The initialized III SDK instance.
    """
    global _instance
    _instance = _create_iii(otel_config)
    return _instance

File: frameworks/motia/motia-js/packages/motia/src/new/iii.ts
import { existsSync, readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import type { InitOptions, ISdk } from 'iii-sdk'
import { registerWorker } from 'iii-sdk'

type OtelConfig = NonNullable<InitOptions['otel']>

const engineWsUrl = process.env.III_URL ?? 'ws://localhost:49134'

function readProjectName(): string | undefined {
  const maxDepth = 1
  let dir = process.cwd()
  for (let i = 0; i <= maxDepth; i++) {
    const pkgPath = join(dir, 'package.json')
    if (existsSync(pkgPath)) {
      try {
        const pkg = JSON.parse(readFileSync(pkgPath, 'utf-8'))
        if (typeof pkg.name === 'string' && pkg.name) {
          return pkg.name
        }
      } catch {
        // ignore
      }
    }
    const parent = dirname(dir)
    if (parent === dir) break
    dir = parent
  }
  return undefined
}

const createIII = (otelConfig?: Partial<OtelConfig>) => {
  return registerWorker(engineWsUrl, {
    otel: {
      enabled: true,
      serviceName: 'motia',
      ...otelConfig,
    },
    telemetry: {
      framework: 'motia',
      project_name: readProjectName(),
    },
  })
}

let instance: ISdk | undefined

export const getInstance = (): ISdk => {
  if (!instance) {
    instance = createIII()
  }
  return instance
}

export const initIII = (otelConfig?: Partial<OtelConfig>) => {
  instance = createIII(otelConfig)
  return instance
}
